src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -219,27 +225,89 @@ private async void DiagnosticService_DiagnosticsUpdated(object sender, Diagnosti
+            // https://docs.microsoft.com/en-us/aspnet/core/blazor/layouts?view=aspnetcore-3.1#centralized-layout-selection
+            // The imported files contents are added to the content of the generated C# file, so we get diagnostics
+            // for both the c# contents in the original razor document and for any of the content in any of the imported files when we query diagnostics for the generated C# file.
+            // These diagnostics will be reported with DiagnosticDataLocation.OriginalFilePath = generated C# file, and DiagnosticDataLocation.MappedFilePath = imported razor file.
+            // This means that in general we could have diagnostics produced by one generated file that map to many different actual razor files.
+            // We can't filter them out as we don't know which razor file(s) the underlying generated C# document actually maps to, and which are just imported.
+            // So we publish them all and let them get de-duped.
+            var fileUriToDiagnostics = diagnostics.GroupBy(diagnostic => GetDiagnosticUri(document, diagnostic)).ToDictionary(
+                group => group.Key,
+                group => group.Select(diagnostic => ConvertToLspDiagnostic(diagnostic, text)).ToArray());
+            return fileUriToDiagnostics;
+
+            static Uri GetDiagnosticUri(Document document, DiagnosticData diagnosticData)
             {
-                Code = diagnostic.Id,
-                Message = diagnostic.Message,
-                Severity = ProtocolConversions.DiagnosticSeverityToLspDiagnositcSeverity(diagnostic.Severity),
-                Range = ProtocolConversions.TextSpanToRange(DiagnosticData.GetExistingOrCalculatedTextSpan(diagnostic.DataLocation, text), text),
-                // Only the unnecessary diagnostic tag is currently supported via LSP.
-                Tags = diagnostic.CustomTags.Contains("Unnecessary") ? new DiagnosticTag[] { DiagnosticTag.Unnecessary } : Array.Empty<DiagnosticTag>()
-            }).ToArray();
+                var filePath = diagnosticData.DataLocation?.MappedFilePath ?? document.FilePath;
+                return ProtocolConversions.GetUriFromFilePath(filePath);
+            }
+
+            static LanguageServer.Protocol.Diagnostic ConvertToLspDiagnostic(DiagnosticData diagnosticData, SourceText text)
+            {
+                return new LanguageServer.Protocol.Diagnostic
+                {
+                    Code = diagnosticData.Id,
+                    Message = diagnosticData.Message,
+                    Severity = ProtocolConversions.DiagnosticSeverityToLspDiagnositcSeverity(diagnosticData.Severity),
+                    Range = GetDiagnosticRange(diagnosticData.DataLocation, text),
+                    // Only the unnecessary diagnostic tag is currently supported via LSP.
+                    Tags = diagnosticData.CustomTags.Contains("Unnecessary") ? new DiagnosticTag[] { DiagnosticTag.Unnecessary } : Array.Empty<DiagnosticTag>()
```

Comment:
1. should we not just pass along all tags?
2. WellKnownDiagnosticTags.Unnecessary

Reply:
LSP only supports the unnecessary tag.  we could pass the others, but I'd prefer to be explicit about what is passed rather than accidentally pass stuff through.

Reply:
Yeah, the last thing we want to do is pass somebody's custom tag that breaks the LSP client. That said, Cyrus's WellKnownDiagnosticTags is good.